# Feed Forward Network and RNN for NLP Tasks

## Feedforward networks for NLP: Classification

The input element $x_i$ could be scalar features like those in Fig. 5.2, e.g., 

- $x_1 = \text{count(words 2 doc)}$, 
- $x_2 = \text{count(positive lexicon words 2 doc)}$, 
- $x_3 = \text{1 if “no” 2 doc}$, and 
- ...  

And the output layer $\hat{y}$ could have 3 nodes (positive, negative, neutral), in which case $\hat{y}_1$ would be the estimated probability of positive sentiment, $\hat{y}_2$ the probability of negative and $\hat{y}_3$ the probability of neutral

The resulting equations
would be just what we saw above for a 2-layer network (as always, we’ll continue
to use the $\sigma$ to stand for any non-linearity, whether sigmoid, ReLU or other)

- $x = [x_1, x_2, ... ,x_N]$ (each $x_i$ is a hand-designed feature)
- $h = \sigma(Wx+b)$
- $z = Uh$
- $\hat{y} = \text{softmax}(z)$

Instead of using hand-built human-engineered features as the input to our classifier, we draw on deep learning’s ability to learn features from the data by representing words as embeddings, like the *word2vec* or *GloVe* embeddings 

For example, for a text with $n$ input words/tokens $w_1, ..., w_n$, we can turn the $n$ embeddings $e(w_1), ..., e(w_n)$ (each of dimensionality $d$) then apply some poolling

- $x = \text{mean}(e(w_1), e(w_2), ... , e(w_n)) $ 
- $h = \sigma(Wx+b)$
- $z = Uh$
- $\hat{y} = \text{softmax}(z)$

The efficience in the computation we need to convert to vector and matrix

- $H = \sigma(XW^T +b)$
- $Z = HU^T$
- $\hat{Y} = \text{softmax}(Z)$

The idea of using word2vec or GloVe embeddings as our input representation— and more generally the idea of relying on another algorithm to have already learned an embedding representation for our input words—is called pretraining

## Feedforward Neural Language Modeling 

Prediction upcoming word from prior word context.

Neural Language Modeling 

Tasks:
- Machine Tranlation
- Text Summarization
- Speech Recognicion
- Grammar Correnction
- Chat 

Compared to *n-gram* models, neural language models can handle much longer histories, can generalize better over contexts of similar words, and are more accurate at word-prediction

On the other hand, neural net language models are much more complex, are slower and need more energy to train, and are less interpretable than n-gram models.

The feedforward neural LM approximates the probability of a word given the entire prior context $P(w_t|w_{1:{t−1}})$ by approximating based on the $N − 1$ previous words

$$ P(w_t|w_1, ..., w_{t−1}) ≈ P(w_t|w_{t−N+1}, ..., w_{t−1})$$

In the following examples we’ll use a 4-gram example, so we’ll show a neural net to estimate the probability $P(w_t = i|w_{t−3};w_{t−2};w_{t−1})$.

Neural language models represent words in this prior context by their embeddings, rather than just by their word identity as used in n-gram language models. Using embeddings allows neural language models to generalize better to unseen data

### Forward inference (decoding) in the neural language model

Forward inference is the task, given an input, of running a forward pass on the
network to produce a probability distribution over possible outputs, in this case next
word

1. Represent $N$ ($N=3$ for our example) previous words in one-hot representation of length $|V|$. 
2. Let $x_{t}$ of the dimention $|V|$ x $1$ `(r, c)`  is one-hot representation of $w_{t}$
3. Given the $E$ matrix of embedding with dimention $\text{d}$ x $\text{|V|}$ (where $d$ is the length of vector that represent the word).
4. Multiply $Ex_{t}=e_t$ of the dimention $\text{d}$ x $1$.
5. For $N=3$ we stacked $e = [e_{t-1}; e_{t-2}; e_{t-3}]$ of the dimention $\text{3d}$ x $1$.
6. Then multiply the weights of the dimention $d_h$ x $\text{3d}$ with stacked embeding, and then apply some non-linear function $h=\sigma(We)$
7. Then $h$ is multiplied by $U$ of the dimention $|V|$ x $d_h$
8. Apply softmax: After the softmax, each node i in the output layer estimates
the probability $P(w_t = i|w_{t−1};w_{t−2};w_{t−3})$

## Recurrent Neural Network

Certain data types such as time-series, text, and biological data contain sequential dependencies. In such cases, a recurrent neural network (RNN) is a type of neural network that can be used to solve such problems. RNN is specifically designed for processing sequential data.


> A Bit of Vector Semantic and Embedding 
> 
> **Embedding**
> 
> It is a way to represent of the meaning of a word in a *vector*. All relate to this word is represented in this vector (semantic, syntax, etc).
> 
> Exists a static embedding and dynamic contextualized embedding (like BERT).
> 
> This avoids the need to feature engineering. 
> 
> This conducts to *self-supervised* ways to learn representations of the input.
> 
> **Vector Semantic**
> 
> It is vector that represents the meaning of a word. Recall, "semantics" is the process of understanding the meaning of a word and this an standar in NLP.
> 
> The idea of vector semantics is to represent a word as a point in a multidimensional space 
> <!-- that is derived from the distributions of embeddings word neighbors. -->
> 
> Vectors for representing words are called *embeddings* (although the term is sometimes more strictly applied only to dense vectors like *word2vec*)
> 
> **Cosine for measuring similarity**
> 
> To measure similarity between two target words v and w, we need a metric that
> takes two vectors gives a measure of their similarity.
> 
> The *dot product* (inner product) acts as a similarity metric because it will tend to be high just when the two vectors have large values in > the same dimensions. Alternatively, vectors that have zeros in different dimensions—orthogonal vectors—will have a dot product of 0, > representing their strong dissimilarity.
> 
> If we normalize this dot product by its length of the vector the result is :
> 
> $$\text{consine} = \frac{\sum_i^N v_iw_i}{\sqrt{\sum_i^N v_i^2}  \sqrt{\sum_i^N w_i^2} } $$
> 
> **Word2vec**

# Sequence Models

#### Motivating Example 

Task: *Entity Recognition*. It is used to find companies's name, times location, currency names, 

$\text{input}$: Harry Potter and Hermione Granger invented a new spell

$$\textbf{x}: [x^{<1>}, x^{<2>} ..., x^{<9>}]$$

$$\textbf{y}: [1, 1, 0, 1, 1, 0, 0, 0, 0]$$

#### Representating Words

$\text{input}$: Harry Potter and Hermione Granger invented a new spell.

We can represent this input sequence in a vector based on this *vocabulary* 

$$\text{vocabulary} = \{\text{a}, \text{aaron}, ...,\text{Harry,} ..., \text{Potter}, ...,\text{zulu}\}$$

We split the sentence in tokens, each token is a word.

$$\text{Input} \: \text{tokens} = [\text{Harry}, \text{and}, \text{Potter}, \text{Hermione}, \text{Granger}, \text{invented}, \text{a}, \text{new}, \text{spell}]$$

$$\text{Input} \: \text{tokens} = [x^{<1>}, x^{<2>}, x^{<3>}, x^{<4>}, x^{<5>}, x^{<6>}, x^{<7>}, x^{<8>}]$$

So $ x^{<1>}=\text{Harry}$ it converted to vector
$$x^{<1>} = \begin{bmatrix} 
0\\ 
0\\
0\\
...\\
1\\
...\\
0\\
...\\
0
\end{bmatrix}$$

This is called one-hot representation. What if we encounter a word that is not in the *vocabulary*? we usually create a token $\text{<UNK>}$.

Other tokes that we use are $\text{<PAD>}$,$\text{<EOS>}$, $\text{<SOS>}$

> Problemns with Feed Forward Neural Network
> * The *inputs* and the *outputs* can have several lenghts. "This is my iPhone" (inputs tokenized by word) (T = 4), "This laptop don't have enough storage" (inputs tokenized by word) (T = 7).
> * The inputs tokenized can be related to earlier inputs tokenized.

## Recurrent Neural Networks

We are focusing on a class of recurrent networks referred to as *Elman Networks* (Elman, 1990) or *simple recurrent networks*.

These networks are useful in their own right and serve as the basis for more
complex approaches like the Long Short-Term Memory (LSTM). In this chapter when we use the term RNN we’ll be referring to
these simpler more constrained networks (although you will often see the term RNN to mean any net with recurrent properties including LSTMs)



A recurrent neural network (RNN) is any network that contains a cycle within its
network connections, meaning that the value of some unit is directly, or indirectly,
dependent on its own earlier outputs as an input.

Critically, this approach does not impose a fixed-length limit on this prior context; the context embodied in the previous hidden layer can include information extending back to the beginning of the sequence.

### Inference in RNNs

Mapping a sequence of inputs to a sequence of outputs

The inference used in this example is for *language modeling task*

Language modeling, or LM, is the use of various statistical and probabilistic techniques to determine the probability of a given sequence of words occurring in a sentence. Language models analyze bodies of text data to provide a basis for their word predictions.

It one of the most basic and important tasks in natural language processing and RNN do very well. 

RNN language models (Mikolov et al., 2010) process the input sequence one
word at a time, attempting to predict the next word from the current word and the
previous hidden state. RNNs thus don’t have the limited context problem that n-gram
models have, or the fixed context that feedforward language models have, since the
hidden state can in principle represent information about all of the preceding words
all the way back to the beginning of the sequence.

$$P(x^{<t+1>} | x^{<1>},...,x^{<t>})$$

In order to diference the input and the output, the next word ($x^{<t+1>}$) is represented by $y^{<t>}$



1. $e^{<t>} = E\textbf{x}^{<t>}$
2. $a^{<t>} = \Phi(W_{aa}a^{<t-1>} + W_{ae}e^{<t>} + b_a)$
3. $y^{<t>} = \Theta(W_{ya}a^{<t>} + b_y)$

> Where $W_{ya}$ are the weights used to compute $y$ multiplied the vector $a$ 
>
> $a^{<0>}$ is a vector of zeros

1. Let $\textbf{x}^{<{t}>}$, of the dimention $|V|$ x $1$, the one-hot representation of $x^{<{t}>}$ (word).
2. Given the $E$ matrix of embedding with dimention ${d}_e$ x $\text{|V|}$ (where $d$ is the length of vector that represent the word).
3. Multiply $E$ with $\textbf{x}^{<t>}$ resulting $e^{<t>}$ of the dimention $d_e$ x $1$.
4. $e^{<t>}$ is multiplied by $W_{ae}$ (of the dimention $d_h$ x $d_e$) and $a^{<t-1>}$ (of the dimention $d_a$ x $1$ with initial values set to zeros) is multiplied by $W_{aa}$ (of the dimention $d_h$ x $d_a$). This last component summatize all preceding relationship.
4. The previous results are sum and added the bias $b_a$ (of dimention $d_h$ x $1$ ), then is passed to no-linear function (ReLu for example) resulting $a^{<t>}$ of dimention $d_h$ x 1.
5. $a^{<t>}$ is multiply by $W_{ya}$ (of dimention $|V|$ x $d_h$) and added a bias $b_y$ (of teh dimention $|V|$ x $1$).
6. At the previously result (called *logits*) apply softmax ($\Theta$) getting the probability distribution over the possible output classes
7. The class that maximize the probability is the class chosen.

<!-- <div><span style="font-size:25px;font-weight:300">RNN Architecture (One Hidden Layer)</span> </div> -->

Graphical representation

<div style = "">
    <div style = "display:flex;align-items:center; justify-content:center; flex-direction:column" >
    <img src = "./assets/architecture-rnn.png" style="height:150px">
    <span>By: Andrew Ng</span>
    </div>
    
</div>


> Note (*Weight tying*)
> 
> This set $d_h=d_e$ in order to use $E$ and $W_{ya}^T$
> 
> *This approach significantly reduces the number of parameters required for the model*

> *Note*
> 
> In actually this is an specific structure called *Many-to-Many* in which the inputs and the outputs has the same length ( $T_x = T_y$) and, also Its is only in one direction (left to right) with one hidden layer.

> In *teacher forcing*
> 
> Instead of using the model's predicted output at each time step as input for the next step, the correct target output (the ground truth) is fed into the model. This means the model gets to see the actual sequence it is supposed to generate during training, which speeds up learning and makes the model converge faster.

#### Examples of RRN architectures

- *Many-to-Many*: Both inputs and outputs are sequences. Can be
direct or delayed.
  - Ex.: Video-captioning, i.e., describing a sequence of images via text (direct).
  - Translating one language into another (delayed)

- *Many-to-One* (Sentimential Classification ) The input data is a sequence, but the output is a fixedsize vector, not a sequence.
  - Ex.: sentiment analysis, the input is some text, and the output is a class label.
- *One-to-Many*:  Input data is in a standard format (not a sequence), theoutput is a sequence.
  - Ex.: Image captioning, where the input is an image, the output is a textdescription of that image

### Training RNN

Back Propagation 

In the Back Propagation we need to compute the loss function.

Let be $\textbf{y}^{<t>}$ (of the dimention $|V|$ x $1$) is the one-hot representation of $y^{<t>}$ (word) and the *logits*  (results before apply softmax of the dimention $|V|$ x $1$).
<!-- {T} -->
1. $L^{<t>} = - {\textbf{y}_i^{<t>}} \log(\text{logits}^{<t>})$ = -$\log(\text{logits}^{<t>})[<t>]$ (from the vector logits extract the index $[<t>]$)
2. $L = \sum_{t=1}^{T_y} L^{<t>}$

<!-- ^{<t>} -->
$$\frac{\partial L}{\partial W_{aa}} = \frac{\partial L^{<t>}}{\partial W_{aa}}_{(t=1)} + \frac{\partial L^{<t>}}{\partial W_{aa}}_{(t=2)} + ... + \frac{\partial L^{<t>}}{\partial W_{aa}}_{(t=T_y)}$$

$$\frac{\partial L}{\partial W_{aa}} = \frac{\partial L^{<t>}}{\partial y^{<t>}} . \Theta\text{'}. W_{ya}. \Phi \text{'}  a^{<t-1>}_{(t=1)} + \frac{\partial L^{<t>}}{\partial y^{<t>}} . \Theta\text{'}. W_{ya}. \Phi \text{'}  a^{<t-1>}[\Phi \text{'}  a^{<t-2>}]_{(t=2)} + ... + \frac{\partial L^{<t>}}{\partial y^{<t>}} . \Theta\text{'}. W_{ya}. \Phi \text{'}  a^{<t-1>}[\Phi \text{'}  a^{<t-2>}]..[\Phi \text{'}  a^{<t-T_y>}]_{(t=T_y)}$$

<!-- $$\frac{\partial L}{\partial W_{aa}} = \frac{\partial L^{<t>}}{\partial y^{<t>}} . \Theta\text{'}. W_{ya}. \Phi \text{'}  a^{<t-1>} + \frac{\partial L^{<t>}}{\partial y^{<t>}} . \Theta\text{'}. W_{ya}. \Phi \text{'}  a^{<t-1>}[\Phi \text{'}  a^{<t-2>}] + ... + \frac{\partial L^{<t>}}{\partial y^{<t>}} . \Theta\text{'}. W_{ya}. \Phi \text{'}  a^{<t-1>}[\Phi \text{'}  a^{<t-2>}]..[\Phi \text{'}  a^{<0>}]$$ -->

This is very problematic:
Vanishing/Exploding gradient problem!

> **Note**
> 
> *Training set*: large corpus (NLP terminology for text with large body) of any language text
> 
> *Tokenize*: Divide the sentences into piece of string (words, n-gram) 
> 
> *Word Tokenize*: divide the sentences into words. Punctuation if is useful can be considere like a word.
> 
> $$\text{cats average 15 hours of sleep a day}.$$
> 
> If considere `.` like a token the tokens would be $y^{<1>}, ..., y^{<8>}, y^{<9>}$, but if it not the tokens would be $y^{<1>}, ..., y^{<8>}$. 
> 
> If we want to tokenize the end of sentence we can add `<EOS>` to tokens.
> 
> If a word is not in a vocabulary (10,000 words) it can be tokenize like `<UNK>` and add to tokens.

*Image generation* and *code generation*, constitute a new area of AI that is often called generative AI

*Autoregressive generation* or *causal LM generation* sample words conditioned on our previous choices. Predicts a value at time $t$ based
on a linear function of the previous values at times $t−1$, $t−2$, and so on. 

Recall, the predicction start with appropriate *context*. It is state-of-the-art for taks like machine translation, summarization, and question answering.

For translation the context is the sentence in the source language; for summarization it’s the long text we want to summarize.

### Stacked and Bidirectional RNN architectures



- Stacked RNNs consist of multiple networks where the output of one layer serves as
the input to a subsequent layer

- A bidirectional RNN (Schuster and Paliwal, 1997) combines two independent RNNs, one where the input is processed from the start to the end, and the other from the end to the start.
Bidirectional RNNs have also proven to be quite effective for sequence classification. 


<div style = "">
    <div style = "display:flex;align-items:center; justify-content:center; flex-direction:column" >
    <img src = "./assets/Bidirectional RNN.png" style="height:180px; border: 0.5px solid;">
    </div>
    
</div>


<!-- ### In the practice

Let the sequence `"The reinforcement learning is the key of machine learning"` and we want to build a *Language Model* to predict the next word. 

So first we need to convert the sequence into *tokens* (in this case the token is a word). 

So the sequence is convert to a vector: $[\text{The}, \: \text{reinforcement}, \: \text{learning},\: \text{is}, \:\text{the}, \: \text{key}, \: \text{of},\: \text{machine},\: \text{learning}]$. 

Above sequence is splitted in *Source Text* and *Target Text* of equal length. $T_x = T_y$

Source Text : $[\text{The}, \: \text{reinforcement}, \: \text{learning},\: \text{is}, \:\text{the}, \: \text{key}, \: \text{of},\: \text{machine}]$. 

Target Text : $[\text{reinforcement}, \: \text{learning},\: \text{is}, \:\text{the}, \: \text{key}, \: \text{of},\: \text{machine},\: \text{learning}]$. 

We can fomalize into variables

Let be $x^{<t>} = [x^{<1>}, .., x^{<8>}]$ and $y^{<t>} = [y^{<1>}, .., y^{<8>}]$ -->

<!-- When $x^{<t>}$ and $y^{<t>}$ is fed to RNN (1 hidden layer) to train, it apply the following steps:

- $x^{<1>} = \text{The}$ is transformed using *embedding transformation* (we define the size of vocabulary and the dimention of embedding) or *one-hot* (based in to the vocabulary). So if $\text{vocabulary}\_\text{size} = 20$ and the dimention of  $\text{embedding}\_\text{size} = 5$ the vector would be $x^{<1>} = [0.79632116, 0.09251072, 0.20748794, 0.20226105, 0.533899253]$ and each time that the word *the* needs to encode this vector will be used.

- $a^{<0>}$ init with zero values $[0, 0, 0, 0, 0]$, $W_{aa}$ and $W_{ax}$ are matrices  of shape = $(5, 5)$ inits with random values. Where the first $5$ to left are the weights for the first unit and $b_a$ is a vector of shape $(1, 5)$ where each single value is the bias for each unit.
- With those parameters and data we can compute $a^{<1>} =\Phi(a^{<0>}\text{.}W_{aa} + x^{<1>}\text{.}W_{ax} + b_a)$.

- Since we need to predict the probability of the next word, we use $\text{softmax}$ function and output should be equal to $20$ the size of the vocabulary. So we can compute the probabilities $P(y^{<1>}), P(y^{<2>}), ..., P(y^{<20>})$.

- With those requeriments the weights $W_{ya}$ is a matrix of shape $(5,20)$ and $b_y$ is vecor of shape $(1, 20)$ with random values at init.

- So we can compute the $y^{<1>} = \Theta(a^{<1>}\text{.}W_{ya} + b_y)$.
- In the next iterartion we will apply the same steps but with the data and parameters of the previous iteration. -->

In [82]:
import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader

In [83]:
# Define a simple dataset for word prediction
class WordDataset(Dataset):
    def __init__(self, sequences, sequence_length):
        
        self.sequences = sequences
        self.sequence_length = sequence_length

    def __len__(self):
        return len(self.sequences)

    def __getitem__(self, idx):
        sequence = self.sequences[idx]
        input_sequence = torch.tensor(sequence[:-1], dtype=torch.long)
        target = torch.tensor(sequence[1:], dtype=torch.long)
        return input_sequence, target

In [ ]:
# Define the RNN model
class RNN(nn.Module):
    def __init__(self, vocabulary_size, embedding_size, hidden_size, output_size):
        super(RNN, self).__init__()
        self.hidden_size = hidden_size
        self.embedding = nn.Embedding(vocabulary_size, embedding_size)
        self.rnn = nn.RNN(embedding_size, hidden_size, batch_first=True)
        self.fc = nn.Linear(hidden_size, output_size)

    def forward(self, x, hidden):
        # x_{(batch_size, sequence_length)}
        # hidden_{(batch_size, hidden_size)}
        embedded = self.embedding(x)
        # embedded_{(batch_size, sequence_length, embedding_size)}
        out, hidden = self.rnn(embedded, hidden)
        # out: Outputs at all time steps.
        # out_{(batch_size, sequence_length, hidden_size)}
        # hidden = out[:, -1]: Final hidden state.
        # hidden_{(batch_size, hidden_size)}
        out = self.fc(out)
        # out_{(batch_size, sequence_length, output_size)}
        return out, hidden

In [85]:
# Function to generate a simple dataset
def generate_dataset():
    # Simple dataset of word sequences
    sequences = [
        [0, 1, 2, 3, 4],  # "hello world"
        [5, 6, 7, 8, 4],  # "how are you"
        [9, 1, 10, 3, 4], # "goodbye world"
        [5, 6, 7, 8, 4],  # "how are you"
    ]
    return sequences

In [86]:
# Define hyperparameters

# input_size = 11
vocabulary_size = 11
# Input size = Vocabulary size
# This implies that the range of posibles values 
# is between 0-10. 
hidden_size = 8 # Size of the hidden state
output_size = 11  # Output size (same as input size)
sequence_length = 4  # Sequence length for training
batch_size = 2  # Batch size for training
num_epochs = 100  # Number of training epochs
embedding_size = 5

# Create the dataset and dataloader
sequences = generate_dataset()
dataset = WordDataset(sequences, sequence_length)
dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=True)

# Create the RNN model instance
model = RNN(vocabulary_size, embedding_size, hidden_size, output_size)

# Define the loss function and optimizer
# For multinomial model
criterion = nn.CrossEntropyLoss()
optimizer = optim.SGD(model.parameters(), lr=0.1)

In [87]:
# Training loop
for epoch in range(num_epochs):
    total_loss = 0
    for idx, (batch_inputs, batch_targets) in enumerate(dataloader):
        # print("Batch inputs: ", batch_inputs)
        optimizer.zero_grad()
        hidden = torch.zeros(1, batch_size, hidden_size) # Initialize hidden state
        outputs, _ = model(batch_inputs, hidden)
        
        loss = criterion(outputs.view(-1, output_size), batch_targets.view(-1))
        loss.backward()
        optimizer.step()
        total_loss += loss.item()    
        print(f'Epoch {epoch+1}/{num_epochs}, Loss: {total_loss/len(dataloader):.4f}')
        # if idx == 0:
        #    break


Epoch 1/100, Loss: 1.2252
Epoch 1/100, Loss: 2.4571
Epoch 2/100, Loss: 1.1886
Epoch 2/100, Loss: 2.3832
Epoch 3/100, Loss: 1.1532
Epoch 3/100, Loss: 2.3115
Epoch 4/100, Loss: 1.1371
Epoch 4/100, Loss: 2.2407
Epoch 5/100, Loss: 1.0836
Epoch 5/100, Loss: 2.1707
Epoch 6/100, Loss: 1.0085
Epoch 6/100, Loss: 2.1184
Epoch 7/100, Loss: 0.9622
Epoch 7/100, Loss: 2.0504
Epoch 8/100, Loss: 0.9982
Epoch 8/100, Loss: 1.9674
Epoch 9/100, Loss: 0.9504
Epoch 9/100, Loss: 1.9026
Epoch 10/100, Loss: 0.9189
Epoch 10/100, Loss: 1.8392
Epoch 11/100, Loss: 0.7914
Epoch 11/100, Loss: 1.7931
Epoch 12/100, Loss: 0.7526
Epoch 12/100, Loss: 1.7327
Epoch 13/100, Loss: 0.7153
Epoch 13/100, Loss: 1.6743
Epoch 14/100, Loss: 0.8006
Epoch 14/100, Loss: 1.6028
Epoch 15/100, Loss: 0.7860
Epoch 15/100, Loss: 1.5490
Epoch 16/100, Loss: 0.7600
Epoch 16/100, Loss: 1.4975
Epoch 17/100, Loss: 0.5828
Epoch 17/100, Loss: 1.4628
Epoch 18/100, Loss: 0.6989
Epoch 18/100, Loss: 1.4020
Epoch 19/100, Loss: 0.6896
Epoch 19/100, Loss:

In [88]:
# Prediction
test_input = torch.tensor([[0, 1, 2, 3]], dtype=torch.long)  # Input sequence: "hell"
hidden = torch.zeros(1, 1, hidden_size)  # Initialize hidden state
with torch.no_grad():
    output, _ = model(test_input, hidden)
predicted_idx = torch.argmax(output, dim=2)
print("Predicted sequence:", [idx.item() for idx in predicted_idx.squeeze()])

Predicted sequence: [1, 2, 3, 4]


## GRU

Links

-  [Gated recurrent unit](https://en.wikipedia.org/wiki/Gated_recurrent_unit)

## The LSTM

The main reason to 

Downsides of using RNN:

- *Retrieval Distant Information* is difficult since the information encoded in hidden states tends to be fairly local, more relevant to the most recent parts of the input sequence and recent decisions. This happen since the weights perform two task simultaneously *provide information useful for the current decision, and updating and carrying forward information required for future decisions*.
- A second difficulty with training RNN is *vanishing gradients problem*.

**LSTM** divide the context management problem into two subproblems: removing information no longer
needed from the context, and adding information likely to be needed for later decision making.

LSTMs accomplish this by first adding an explicit context layer to the architecture (in addition to the
usual recurrent hidden layer), and through the use of specialized neural units that
make use of gates to control the flow of information into and out of the units that
comprise the network layers. 

The *gates* in an LSTM share a common design pattern; each consists of a feedforward layer, followed by a sigmoid activation function, followed by a pointwise multiplication with the layer being gated.


Values in the layer being *gated* that align with values near $1$ in the mask are passed through
nearly unchanged; values corresponding to lower values are essentially erased

The first gate we’ll consider is the **forget gate**. The purpose of this gate is to delete information from the context that is no longer needed.
- $f_t = \sigma(U_f h_{t-1} + W_f x_t)$
- $k_t = c_{t-1} \odot f_{t}$

The next task is to compute the actual information we need to extract from the previous hidden state and current inputs—the same basic computation we’ve been using for all our recurrent networks

- $g_t=\tanh(U_g h_{t-1} + W_g x_{t})$

Next, we generate the mask for the add gate to select the information to add to the *current context*

- $i_t = \sigma(U_i h_{t-1} + W_i x_{t})$
- $j_t = g_t \odot i_t$

Next, we add this to the modified context vector to get our new context vector

- $c_t = k_t + j_t $

The final gate we’ll use is the output gate which is used to decide what information is required for the current hidden state (as opposed to what information needs to be preserved for future decisions
- $o_t=\sigma(U_o h_{t-1} + W_o x_{t})$
- $h_t = o_t \odot \tanh(c_t) $

## The Encoder-Decoder Model with RNN

The *encoder-decoder* (sequence-to-sequence) networks model is used in *machine translation task*, in which the input and output sequence has different lenght.

The key idea is train an *encoder network* that take an input sequence and create a context that is passed to *decoder network* that generate a sequence.

1. An *encoder* that accepts an input sequence, $x$, and generates a corresponding
sequence of contextualized representations, $h^e$ . LSTMs, convolutional networks, and Transformers can all be employed as encoders.
2. A *context vector*, $c$, which is a function of $h^e$, and conveys (carry) the essence (of information) of the input to the decoder.
3. A *decoder*, which accepts $c$ as input and generates an arbitrary length sequence of hidden states $h^d$ , from which a corresponding sequence of output states $y$ , can be obtained. Just as with encoders, decoders can be realized by
any kind of sequence architecture

<div style = "">
    <div style = "display:flex;align-items:center; justify-content:center; flex-direction:column" >
    <img src = "./assets/encoder-decoder-model.png" style="height:220px;">
    </div>
    
</div>


> *Note*
> 
> The entire purpose of the encoder is to generate a contextualized representation of the input

0. $e^{<t>}_x = E\textbf{x}^{<t>}$
1. $a^{<t>}_e = \Phi(W_{aa}a^{<t-1>}_e + W_{ae}e^{<t>}_x + b_a)$
2. $c = a^{T_x}_e=a^{0}_d$, $y^{0} = \text{<s>}$
3. $e^{<\tau>}_y = E\textbf{y}^{<\tau>}$
4. $a^{<\tau>}_d = \Phi(W_{aa}a^{<\tau-1>}_d + W_{ae}e^{<\tau>}_y + W_{ac}c + b_a)$
<!-- 5. $c = a^{T_x}_e$ -->
5. $y^{<\tau>} = \Theta(W_{ya}a^{<\tau>}_d + b_y)$

> Where $W_{ya}$ are the weights used to compute $y$ multiplied the vector $a$ 
>
> $a^{<0>}$ is a vector of zeros

Enconder 
1. Let $\textbf{x}^{<{t}>}$, of the dimention $|V|$ x $1$, the one-hot representation of $x^{<{t}>}$ (word).
2. Given the $E$ matrix of embedding with dimention ${d}_e$ x $\text{|V|}$ (where $d$ is the length of vector that represent the word).
3. Multiply $E$ with $\textbf{x}^{<t>}$ resulting $e^{<t>}_x$ of the dimention $d_e$ x $1$.
4. $e^{<t>}_x$ is multiplied by $W_{ae}$ (of the dimention $d_h$ x $d_e$) and $a^{<t-1>}_x$ (of the dimention $d_a$ x $1$ with initial values set to zeros) is multiplied by $W_{aa}$ (of the dimention $d_h$ x $d_a$). This last component summarize all preceding relationship.
4. The previous results are sum and added the bias $b_a$ (of dimention $d_h$ x $1$ ), then is passed to no-linear function (ReLu for example) resulting $a^{<t>}_e$ of dimention $d_h$ x 1.
5. At time $t=T_x$ (the length of the input sequence) we obtain $a^{T_x}_e$, this is the context vector ($c$) used in decoder net, the vector that summarize all input information.

Decoder

1. Start with initial values $a^{<0>}_d=c$, $y^{<0>}=\text{<s>}$
1. Let $\textbf{y}^{<\tau>}$, of the dimention $|V|$ x $1$, the one-hot representation of $y^{<\tau>}$ (word).
2. Given the $E$ matrix of embedding with dimention ${d}_e$ x $\text{|V|}$ (where $d$ is the length of vector that represent the word).
3. Multiply $E$ with $\textbf{y}^{<\tau>}$ resulting $e^{<\tau>}_y$ of the dimention $d_e$ x $1$.
4. $e^{<\tau>}_y$ is multiplied by $W_{ae}$ (of the dimention $d_h$ x $d_e$), $c$ (of the dimention $d_a$ x $1$) is multiplied by $W_{ac}$ (of the dimention $d_h$ x $d_a$) and $a^{<\tau-1>}_d$ (of the dimention $d_a$ x $1$ with initial values set to zeros) is multiplied by $W_{aa}$ (of the dimention $d_h$ x $d_a$). 
4. The previous results are sum and added the bias $b_a$ (of dimention $d_h$ x $1$ ), then is passed to no-linear function (ReLu for example) resulting $a^{<\tau>}_d$ of dimention $d_h$ x 1.
5. $a^{<\tau>}_d$ is multiply by $W_{ya}$ (of dimention $|V|$ x $d_h$) and added a bias $b_y$ (of the dimention $|V|$ x $1$).
6. At the previously result (called *logits*) apply softmax ($\Theta$) getting the probability distribution over the possible output classes
7. The class that maximize the probability is the class chosen.

In training, therefore, it is more common to use teacher forcing in the
decoder. *Teacher forcing* means that we force the system to use the gold target token
from training as the next input, rather than allowing it to rely on the (possibly
erroneous) decoder output.

## Attention 

Loung Attention 

Downsides of using a simple encoder-decoder:
- The final hidden state of the econder is thus acting as a **bottleneck**. It must represent absolutely
everything about the meaning of the source text, since the only thing the decoder
knows about the source text is what’s in this context vector

The attention mechanism is a solution to the bottleneck problem, a way of
allowing the decoder to get information from all the hidden states of the encoder,
not just the last hidden state

The idea of attention is instead to create the single fixed-length vector c by taking
a weighted sum of all the encoder hidden states

The weights focus on (‘attend to’) a particular part of the source text that is relevant for the token the decoder is
currently producing

Attention thus replaces the static context vector with one that
is dynamically derived from the encoder hidden states

This is the weighted average over all the *encoder hidden states*.

So $c \rightarrow c^{<\tau>} = \sum_{<t>=1}^{T_x} \alpha^{<t,\tau>} a^{<t>}_e$

Where the alpha tells us the proportional relevance of each encoder hidden
state $<t>$ to the prior hidden decoder state $a^{<\tau>}_d$

$\alpha^{<t,\tau>} = \text{softmax}(\text{score}(a^{<t>}_e, a^{<\tau>}_d))$



Dot Product Attenction

$\text{score}(a^{<t>}_e, a^{<\tau>}_d) = a^{<\tau>}_d.a^{<t>}_e$

Implements relevance as attention similarity: measuring how similar the decoder hidden state is to an encoder hidden state

🧠 Intuition
Imagine you're translating a sentence, and the current output token you're trying to predict is likely to be a verb. The decoder’s hidden state reflects that context — say, it's expecting a verb.

The attention mechanism will search through the encoder’s hidden states, and those that encode verbs or verb-related meanings will look more similar to the decoder’s hidden state — so they get higher weights.

## Application

The most common way for checking if an email is valid or not is with authentication. But if the company where you work not have an application that can check if an email is valid or not, then you can build a model that can correct the email (correct only domain part). 

Parts of an email:

Email example: $\text{support} @ \text{getwemail.com}$
- `support`: username
- `@`: symbol
- `getwemail`: mail server
- `com`: domain

$$\text{wings@gmial.com} \rightarrow \text{wings@gmail.com}$$

We will be applying the following concepts:
- Embedding
- RNN (GRU or LSTM)

In [21]:
from __future__ import unicode_literals, print_function, division
from io import open
import re
import torch
import unicodedata
import torch.nn as nn
import matplotlib.pyplot as plt
# plt.switch_backend('agg')
import matplotlib.ticker as ticker
import random
from torch import optim
import torch.nn.functional as F
import numpy as np
import pandas as pd
import time, os, glob
from torch.utils.data import TensorDataset, DataLoader, RandomSampler
import string
import unicodedata

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

In [3]:
vowels = "aeiou"
consonants = ''.join([x for x in string.ascii_lowercase if x not in vowels])
tilded_vowels = "áéíóú"
special_characters = "ñç"
# End of line token
EOL = '<'
SOL = '>'
MAX_LENGTH = 40

In [4]:
# We can use "_" to represent an out-of-vocabulary character, that is, any character we are not handling in our model
allowed_characters =  (SOL+string.digits + vowels + tilded_vowels + 
                       consonants + special_characters + ".,;'-_"+EOL)

n_letters = len(allowed_characters)
EOL_token = n_letters - 1
SOL_token = 0

In [5]:
# Find letter index from all_letters, e.g. "a" = 0
def letterToIndex(letter):
    # return our out-of-vocabulary character if we encounter a letter unknown to our model
    if letter not in allowed_characters:
        return allowed_characters.find("_")
    else:
        return allowed_characters.find(letter)

# Turn a line into a <line_length x 1 x n_letters>,
# or an array of one-hot letter vectors
def lineToTensor(line):
    tensor = torch.zeros(len(line), 1, n_letters)
    for li, letter in enumerate(line):
        tensor[li][0][letterToIndex(letter)] = 1
    return tensor

def indexesFromLine(line):
    return [letterToIndex(li) for li in line]

def tensorFromLine(line):
    indexes = indexesFromLine(line)
    # indexes.append(EOS_token)
    return torch.tensor(indexes, dtype=torch.long, device=device).view(1, -1)

In [6]:
def readEmail(path_file):
    print("Reading lines...")

    # Read the file and split into lines

    with open(path_file, encoding='utf-8') as f:
            next(f) #skip the first line
            lines = f.read().strip().split('\n')
            
    # Split every line into pairs and normalize
    pairs = [[s for s in l.split(',')] for l in lines]

    return pairs

In [7]:
def prepareData(path_file):
    pairs = readEmail(path_file)
    print("Read %s sentence pairs" % len(pairs))
    return pairs

pairs = prepareData("data_model_email.csv")
print(random.choice(pairs))

Reading lines...
Read 705 sentence pairs
['eyg.edu.pe', 'eyg.edu.pe']


In [8]:
def get_dataloader(batch_size):
    pairs = prepareData("data_model_email.csv")

    n = len(pairs)
    input_ids = np.zeros((n, MAX_LENGTH), dtype=np.int32)
    target_ids = np.zeros((n, MAX_LENGTH), dtype=np.int32)

    for idx, (inp, tgt) in enumerate(pairs):
        inp_ids = indexesFromLine(inp)
        tgt_ids = indexesFromLine(tgt)
        inp_ids.append(EOL_token)
        tgt_ids.append(EOL_token)
        input_ids[idx, :len(inp_ids)] = inp_ids
        target_ids[idx, :len(tgt_ids)] = tgt_ids

    train_data = TensorDataset(torch.LongTensor(input_ids).to(device),
                               torch.LongTensor(target_ids).to(device))

    train_sampler = RandomSampler(train_data)
    train_dataloader = DataLoader(train_data, sampler=train_sampler, batch_size=batch_size)
    return train_dataloader

In [9]:
class EncoderRNN(nn.Module):
    def __init__(self, input_size, hidden_size, dropout_p=0.1):
        super(EncoderRNN, self).__init__()
        self.hidden_size = hidden_size

        self.embedding = nn.Embedding(input_size, hidden_size)
        self.gru = nn.GRU(hidden_size, hidden_size, batch_first=True)
        self.dropout = nn.Dropout(dropout_p)

    def forward(self, input):
        embedded = self.dropout(self.embedding(input))
        # hidden Defaults to zeros if not provided to GRU
        output, hidden = self.gru(embedded)
        return output, hidden

In [10]:
class DecoderRNN(nn.Module):
    def __init__(self, hidden_size, output_size):
        super(DecoderRNN, self).__init__()
        self.embedding = nn.Embedding(output_size, hidden_size)
        self.gru = nn.GRU(hidden_size, hidden_size, batch_first=True)
        self.out = nn.Linear(hidden_size, output_size)

    def forward(self, encoder_outputs, encoder_hidden, target_tensor=None):
        batch_size = encoder_outputs.size(0)
        decoder_input = torch.empty(batch_size, 1, dtype=torch.long, device=device).fill_(SOL_token)
        decoder_hidden = encoder_hidden
        decoder_outputs = []

        for i in range(MAX_LENGTH):
            decoder_output, decoder_hidden  = self.forward_step(decoder_input, decoder_hidden)
            decoder_outputs.append(decoder_output)

            if target_tensor is not None:
                # Teacher forcing: Feed the target as the next input
                decoder_input = target_tensor[:, i].unsqueeze(1) # Teacher forcing
            else:
                # Without teacher forcing: use its own predictions as the next input
                _, topi = decoder_output.topk(1)
                decoder_input = topi.squeeze(-1).detach()  # detach from history as input

        decoder_outputs = torch.cat(decoder_outputs, dim=1)
        decoder_outputs = F.log_softmax(decoder_outputs, dim=-1)
        # We return `None` for consistency in the training loop
        # in attection models, the outup includes the attention weights
        # but we don't need them here
        return decoder_outputs, decoder_hidden, None 

    def forward_step(self, input, hidden):
        # x{t}: batch_size x 1: 
        output = self.embedding(input)
        # Internally in embedding
        # X{t}:batch_size x  1 x n_letters (one-hot)
        # E:n_letters x embedding_size (embedding matrix)
        # e{t}: batch_size x  1 x embedding_size
        output = F.relu(output)
        # e{t}=relu(e):batch_size x  1 x embedding_size
        # a{t-1}: batch_size x  1 x embedding_size
        # `hidden` defaults to zeros if not provided
        output, hidden = self.gru(output, hidden)
        # W{e,a}:embedding_size x hidden_size
        # a{t} = a{t-1} x W{a,a} + e{t} x W{e,a} + b{t}:batch_size x  1 x hidden_size
        output = self.out(output)
        # W{a,y}:hidden_size x output_size
        # y{t} = a{t} x W{a,y}:batch_size x  1 x output_size
        return output, hidden

In [11]:
class BahdanauAttention(nn.Module):
    def __init__(self, hidden_size):
        super(BahdanauAttention, self).__init__()
        self.Wa = nn.Linear(hidden_size, hidden_size)
        self.Ua = nn.Linear(hidden_size, hidden_size)
        self.Va = nn.Linear(hidden_size, 1)

    def forward(self, query, keys):
        scores = self.Va(torch.tanh(self.Wa(query) + self.Ua(keys)))
        scores = scores.squeeze(2).unsqueeze(1)

        weights = F.softmax(scores, dim=-1)
        context = torch.bmm(weights, keys)

        return context, weights

class AttnDecoderRNN(nn.Module):
    def __init__(self, hidden_size, output_size, dropout_p=0.1):
        super(AttnDecoderRNN, self).__init__()
        self.embedding = nn.Embedding(output_size, hidden_size)
        self.attention = BahdanauAttention(hidden_size)
        self.gru = nn.GRU(2 * hidden_size, hidden_size, batch_first=True)
        self.out = nn.Linear(hidden_size, output_size)
        self.dropout = nn.Dropout(dropout_p)

    def forward(self, encoder_outputs, encoder_hidden, target_tensor=None):
        batch_size = encoder_outputs.size(0)
        decoder_input = torch.empty(batch_size, 1, 
                                    dtype=torch.long, device=device).fill_(SOL_token)
        decoder_hidden = encoder_hidden
        decoder_outputs = []
        attentions = []

        for i in range(MAX_LENGTH):
            decoder_output, decoder_hidden, attn_weights = self.forward_step(
                decoder_input, decoder_hidden, encoder_outputs
            )
            decoder_outputs.append(decoder_output)
            attentions.append(attn_weights)

            if target_tensor is not None:
                # Teacher forcing: Feed the target as the next input
                decoder_input = target_tensor[:, i].unsqueeze(1) # Teacher forcing
            else:
                # Without teacher forcing: use its own predictions as the next input
                _, topi = decoder_output.topk(1)
                decoder_input = topi.squeeze(-1).detach()  # detach from history as input

        decoder_outputs = torch.cat(decoder_outputs, dim=1)
        decoder_outputs = F.log_softmax(decoder_outputs, dim=-1)
        attentions = torch.cat(attentions, dim=1)

        return decoder_outputs, decoder_hidden, attentions


    def forward_step(self, input, hidden, encoder_outputs):
        embedded =  self.dropout(self.embedding(input))

        query = hidden.permute(1, 0, 2)
        context, attn_weights = self.attention(query, encoder_outputs)
        input_gru = torch.cat((embedded, context), dim=2)

        output, hidden = self.gru(input_gru, hidden)
        output = self.out(output)

        return output, hidden, attn_weights

In [12]:
def train_epoch(dataloader, encoder, decoder, encoder_optimizer,
          decoder_optimizer, criterion):

    total_loss = 0
    for data in dataloader:
        input_tensor, target_tensor = data

        encoder_optimizer.zero_grad()
        decoder_optimizer.zero_grad()

        encoder_outputs, encoder_hidden = encoder(input_tensor)
        decoder_outputs, _, _ = decoder(encoder_outputs, encoder_hidden, target_tensor)

        loss = criterion(
            decoder_outputs.view(-1, decoder_outputs.size(-1)),
            target_tensor.view(-1)
        )
        loss.backward()

        encoder_optimizer.step()
        decoder_optimizer.step()

        total_loss += loss.item()

    return total_loss / len(dataloader)

In [13]:
import time
import math

def asMinutes(s):
    m = math.floor(s / 60)
    s -= m * 60
    return '%dm %ds' % (m, s)

def timeSince(since, percent):
    now = time.time()
    s = now - since
    es = s / (percent)
    rs = es - s
    return '%s (- %s)' % (asMinutes(s), asMinutes(rs))

In [ ]:
def showPlot(points):
    # plt.figure()
    _, ax = plt.subplots()
    ax.plot(range(1, len(points)+1),points)
    # this locator puts ticks at regular intervals
    loc = ticker.MultipleLocator(base=0.2)
    ax.yaxis.set_major_locator(loc)
    # plt.plot(points)

In [42]:
def train(train_dataloader, encoder, decoder, n_epochs, learning_rate=0.001,
               print_every=100, plot_every=100, min_loss=0.0006):
    start = time.time()
    plot_losses = []
    print_loss_total = 0  # Reset every print_every
    plot_loss_total = 0  # Reset every plot_every
    loss_total = []

    encoder_optimizer = optim.Adam(encoder.parameters(), lr=learning_rate)
    decoder_optimizer = optim.Adam(decoder.parameters(), lr=learning_rate)
    criterion = nn.NLLLoss()

    for epoch in range(1, n_epochs + 1):
        loss = train_epoch(train_dataloader, encoder, decoder, encoder_optimizer, decoder_optimizer, criterion)
        print_loss_total += loss
        plot_loss_total += loss
        loss_total.append(loss)

        if epoch % print_every == 0:
            print_loss_avg = print_loss_total / print_every
            print_loss_total = 0
            print('%s (%d %d%%) %.4f' % (timeSince(start, epoch / n_epochs),
                                        epoch, epoch / n_epochs * 100, print_loss_avg))

        if epoch % plot_every == 0:
            plot_loss_avg = plot_loss_total / plot_every
            plot_losses.append(plot_loss_avg)
            plot_loss_total = 0
        
        if  sum(loss_total)/len(loss_total)<min_loss:
            break    

    showPlot(plot_losses)

In [43]:
def evaluate(encoder, decoder, line):
    with torch.no_grad():
        input_tensor = tensorFromLine(line)

        encoder_outputs, encoder_hidden = encoder(input_tensor)
        decoder_outputs, decoder_hidden, decoder_attn = decoder(encoder_outputs, encoder_hidden)

        _, topi = decoder_outputs.topk(1)
        decoded_ids = topi.squeeze()

        decoded_words = []
        for idx in decoded_ids:
            if idx.item() == EOL_token:
                decoded_words.append('<EOS>')
                break
            decoded_words.append(allowed_characters[idx.item()])
    return decoded_words, decoder_attn

In [28]:
def evaluateRandomly(encoder, decoder, n=10):
    for i in range(n):
        pair = random.choice(pairs)
        print('>', pair[0])
        print('=', pair[1])
        output_words, _ = evaluate(encoder, decoder, pair[0])
        output_sentence = ''.join(output_words)
        print('<', output_sentence)
        print('')

Define the parameters

In [44]:
hidden_size = 128
batch_size = 20
n_epochs = 100
min_loss = 0.0006

In [21]:
train_dataloader = get_dataloader(batch_size)

Reading lines...
Read 705 sentence pairs


In [22]:
encoder = EncoderRNN(n_letters, hidden_size).to(device)

Sequence To Sequence Model: Encoder(GRU) To Decoder(GRU)


In [19]:
# decoder = DecoderRNN(hidden_size, n_letters).to(device)
# train(train_dataloader, encoder, decoder, n_epochs, print_every=5, plot_every=5)

Sequence To Sequence Model: Encoder(GRU) To Attention Decoder(GRU)

In [23]:
decoder = AttnDecoderRNN(hidden_size, n_letters).to(device)

train(train_dataloader, encoder, decoder, n_epochs, print_every=5, plot_every=5)

0m 42s (- 13m 24s) (5 5%) 0.8888
1m 25s (- 12m 45s) (10 10%) 0.5551
2m 24s (- 13m 40s) (15 15%) 0.3430
3m 22s (- 13m 31s) (20 20%) 0.1527
4m 6s (- 12m 18s) (25 25%) 0.0738
4m 49s (- 11m 15s) (30 30%) 0.0426
5m 34s (- 10m 21s) (35 35%) 0.0216
6m 17s (- 9m 26s) (40 40%) 0.0432
7m 2s (- 8m 35s) (45 45%) 0.0084
7m 45s (- 7m 45s) (50 50%) 0.0054
8m 31s (- 6m 58s) (55 55%) 0.0041
9m 14s (- 6m 9s) (60 60%) 0.0034
10m 0s (- 5m 23s) (65 65%) 0.0125
10m 50s (- 4m 38s) (70 70%) 0.0092
11m 46s (- 3m 55s) (75 75%) 0.0042
12m 57s (- 3m 14s) (80 80%) 0.0014
14m 7s (- 2m 29s) (85 85%) 0.0010
15m 17s (- 1m 41s) (90 90%) 0.0008
16m 28s (- 0m 52s) (95 95%) 0.0006
17m 26s (- 0m 0s) (100 100%) 0.0008


In [45]:
encoder.eval()
decoder.eval()
evaluateRandomly(encoder, decoder)

> ic.edu.pe
= ic.edu.pe
< ic.edu.pe<EOS>

> ilac.edu.pe
= ilac.edu.pe
< ilac.edu.edu.pe<EOS>

> ramirez.com.pe
= ramirez.com.pe
< ramirez.com.pe<EOS>

> stakeholders.com.pe
= stakeholders.com.pe
< stakeholders.com.pe<EOS>

> hotmamil.com
= hotmail.com
< hotmail.com<EOS>

> cosola.com
= cosola.com
< osla.com<EOS>

> auditorescorporativos.com
= auditorescorporativos.com
< idtorescorporativos.com<EOS>

> softmediaconsultores.com
= softmediaconsultores.com
< stemiaconsultores.com<EOS>

> cusco.coar.edu.pe
= cusco.coar.edu.pe
< cusoc.coar.edu.pe<EOS>

> entel.cl
= entel.cl
< entl.clcl<EOS>



In [ ]:
def showAttention(line, output_words, attentions):
    # print(attentions.cpu().numpy().shape)
    # print(line)
    # print(output_words)
    fig, ax = plt.subplots()
    # ax = fig.add_subplot(111)
    cax = ax.matshow(attentions.cpu().numpy(), cmap='bone')
    plt.show()
    # fig.colorbar(cax)

    # Set up axes
    # ax.set_xticklabels([''] + line.split('') +
    #                     ['<EOS>'], rotation=90)
    # ax.set_yticklabels([''] + output_words)
# 
    # # Show label at every tick
    # ax.xaxis.set_major_locator(ticker.MultipleLocator(1))
    # ax.yaxis.set_major_locator(ticker.MultipleLocator(1))

# plt.show()


def evaluateAndShowAttention(line):
    output_words, attentions = evaluate(encoder, decoder, line)
    print('input =', line)
    print('output =', ''.join(output_words))
    showAttention(line, output_words, attentions[0, :len(output_words), :])


evaluateAndShowAttention('gomail.com')

# evaluateAndShowAttention('je suis trop fatigue pour conduire')
# 
# evaluateAndShowAttention('je suis desole si c est une question idiote')
# 
# evaluateAndShowAttention('je suis reellement fiere de vous')

input = gomail.com
output = gmail.com<EOS>


C:\Users\dante\AppData\Local\Temp\ipykernel_35200\2792693239.py:5: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`). Consider using `matplotlib.pyplot.close()`.
  fig, ax = plt.subplots()
C:\Users\dante\AppData\Local\Temp\ipykernel_35200\2792693239.py:8: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


## Transformers

Transformer can handle distant information, but it is not based on recurrent connections


https://cschranz.medium.com/connect-vs-code-to-your-remote-gpu-jupyter-instance-58b86a195d9e

<div style = "display:flex;align-items:center; justify-content:left; flex-direction:row">
 <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-youtube" viewBox="0 0 16 16">
  <path d="M8.051 1.999h.089c.822.003 4.987.033 6.11.335a2.01 2.01 0 0 1 1.415 1.42c.101.38.172.883.22 1.402l.01.104.022.26.008.104c.065.914.073 1.77.074 1.957v.075c-.001.194-.01 1.108-.082 2.06l-.008.105-.009.104c-.05.572-.124 1.14-.235 1.558a2.01 2.01 0 0 1-1.415 1.42c-1.16.312-5.569.334-6.18.335h-.142c-.309 0-1.587-.006-2.927-.052l-.17-.006-.087-.004-.171-.007-.171-.007c-1.11-.049-2.167-.128-2.654-.26a2.01 2.01 0 0 1-1.415-1.419c-.111-.417-.185-.986-.235-1.558L.09 9.82l-.008-.104A31 31 0 0 1 0 7.68v-.123c.002-.215.01-.958.064-1.778l.007-.103.003-.052.008-.104.022-.26.01-.104c.048-.519.119-1.023.22-1.402a2.01 2.01 0 0 1 1.415-1.42c.487-.13 1.544-.21 2.654-.26l.17-.007.172-.006.086-.003.171-.007A100 100 0 0 1 7.858 2zM6.4 5.209v4.818l4.157-2.408z"/>
</svg><span style="padding-left:5px"> Youtube and Links</span></div>


- [Pytorch Transformers from Scratch (Attention is all you need)](https://www.youtube.com/watch?v=U0s0f995w14)
- [Transformer Model Tutorial in PyTorch: From Theory to Code](https://www.datacamp.com/tutorial/building-a-transformer-with-py-torch)
- [Let's build GPT: from scratch, in code, spelled out.](https://www.youtube.com/watch?v=kCc8FmEb1nY)
- [3blue1brown](https://www.youtube.com/@3blue1brown/videos)